vivisect/analysis/ppc/emulation.py

```diff
@@ -9,82 +13,204 @@
+            vw.addVaSet("ppc_spr_writes", (("funcva", viv_const.VASET_ADDRESS), ("writes", viv_const.VASET_COMPLEX)))
+
+        if self.spr_reads and vw.getVaSet('ppc_spr_reads'):
+            vw.setVaSetRow('ppc_spr_reads', (self.fva, dict((k, sorted(v)) for k, v in self.spr_reads.items())))
+        if self.spr_writes and vw.getVaSet('ppc_spr_writes'):
+            vw.setVaSetRow('ppc_spr_writes', (self.fva, dict((k, sorted(v)) for k, v in self.spr_writes.items())))
+
+        # If there was an IVPR write, get any IVORx SPRs written in this
+        # function and log these as possible exception handlers, if the target
+        # addresses register as a valid function make an XREF from the
+        #   mtspr IVOR?, r?
+        # instruction to the exception handler
+        #
+        # We aren't checking code paths here so it's a little stupid, but it's
+        # less trouble this way because it doesn't cause weird code path
+        # deviations.
+        for _, ivpr_value in self.spr_writes.get('IVPR', []):
+            base = ivpr_value & IVPR_MASK
+
+            for ivor_name, namefmt in IVOR_MAP.items():
+                for va, ivor_value in self.spr_writes.get(ivor_name, []):
+                    eva = base | (ivor_value & IVORx_MASK)
+                    logger.debug('Looking for possible exception handler @ 0x%x', eva)
+                    if vw.isProbablyCode(eva):
+                        vw.addEntryPoint(eva)
+                        funcname = namefmt % eva
+                        vw.makeName(eva, funcname)
+                        vw.setComment(va, funcname)
+
+    def checkAddDataXref(self, vw, va, val, discrete):
+        # For PPC data XREFS are only added through load/store instructions
+        # which are tracked using the logread/logwrite feature
+        pass
 
-        viv_monitor.AnalysisMonitor.prehook(self, emu, op, starteip)
 
 def ppcname(idx):
-    ret = ppcargnames.get(idx)
-    if ret == None:
-        name = 'arg%d' % idx
+    if idx >= len(ppcargnames):
+        name = 'arg%d' % (len(ppcargnames) + idx)
     else:
-        name, idx = ret
+        name = ppcargnames[idx][0]
     return name
 
+
 def buildFunctionApi(vw, fva, emu, emumon):
-    
+
     argc = 0
     funcargs = []
     callconv = vw.getMeta('DefaultCall')
-    argnames = ppcargnames
     undefregs = set(emu.getUninitRegUse())
 
-    for argnum in range(len(argnames), 0, -1):
-        argname, argid = argnames[argnum-1]
-        if argid in undefregs:
+    for argnum, (_, argreg) in enumerate(ppcargnames):
+        if argreg not in undefregs:
             argc = argnum
             break
 
-    if emumon.stackmax > 0:
-        targc = (emumon.stackmax // 8) + 6
-        if targc > 40:
+    # Typical PPC stack will store the LR @ 4(<original r1>), any stack
+    # parameters will be loaded from offsets higher than that.
+    if emumon.stackmax > 4:
+        stack_args = (emumon.stackmax-4) // emu.getPointerSize()
```

Comment:
It comes from my interpretation of the PPC elf64abi documentation (https://refspecs.linuxfoundation.org/ELF/ppc64/PPC-elf64abi.html#DYNAM-STACK) and my understanding of how the `buildFunctionApi` function works. However I am not 100% sure.

I am fairly certain I tested this on a real target and got good results with these values, but I can't find my notes on it.

Reply:
32-bit or 64-bit?  this hard-coding requires documentation (and comments)... and i'm still not sure what's going on here.  can you walk me through it tomorrow?

Reply:
I changed the code to use `emu.getPointerSize()` instead of a hardcoded 4 bytes. Both the PPC32 and PPC64 ABIs should be the same in where the parameters are placed on the stack if the pointer size is used.